**심층 신경망**

In [5]:
from sklearn.model_selection import train_test_split
from tensorflow.keras.datasets import fashion_mnist

# 데이터 로드
(train_input, train_target), (test_input, test_target) = fashion_mnist.load_data()

# 스케일링 및 변환
train_scaled = train_input / 255.0  # 픽셀값을 0~1로 스케일링
train_scaled = train_scaled.reshape(-1, 28 * 28)  # (28, 28) -> (784,)

# train_target과 동기화된 상태에서 train_test_split 사용
train_scaled, val_scaled, train_target, val_target = train_test_split(
    train_scaled, train_target, test_size=0.2, random_state=42
)

첫번째 이미지를 펼친 배열 -> 입력층 784개의 뉴런 -> 은닉층 100개의 뉴런 -> 출력층 10개의 뉴런 -> softmax 함수 -> 출력

은닉층: 입력층과 출력층 사이에 있는 모든 층, ex) 활성화 함수 - 시그모이드, relu 함수

**방법1**

In [6]:
dense1=keras.layers.Dense(100, activation='sigmoid', input_shape=(784,)) #keras 신경망의 첫번째 층에서는 항상 input_shape 명시하기 / 100: 뉴런 개수
dense2=keras.layers.Dense(10 ,activation='sigmoid')

NameError: name 'keras' is not defined

In [ ]:
model=keras.Sequential([dense1, dense2])

In [ ]:
model.summary()

Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 100)            │        78,500 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 10)             │         1,010 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 79,510 (310.59 KB)

 Trainable params: 79,510 (310.59 KB)

 Non-trainable params: 0 (0.00 B)

1. (None, 100): 샘플의 개수가 None

keras의 fit()은 train data를 여러 번 나누어 경사 하강법을 수행함 (미니배치 경사 하강법) 기본 미니배치 크기는 32개이다 (fit에서 batch_size 매개변수로 바꿀 수 있음) 

샘플 개수에 고정되지 않고 어떤 batch에도 유연하게 대응하기 위해 None으로 설정함

샘플마다 784개의 픽셀값이 은닉층을 통과하고 100개의 특성으로 압축됨

2. Param: Dense 층이므로 픽셀 784개와 100개의 조합 + 100개의 절편

3. 두번째 Param: 1010임, 100개의 은닉층 뉴런과 10개의 출력층 뉴런을 연결 (1000) + 10개 절편

4. Non-trainable params: 훈련되지 않는 파라미터 ex) 경사하강법으로 훈련되지 않는 parameter를 가진 층

**방법2**

In [ ]:
model = keras.Sequential([
    keras.layers.Dense(100, activation='sigmoid', input_shape=(784,), name='hidden'),
    keras.layers.Dense(10, activation='softmax', name = 'output')
], name='패션 Mnist 모델')

c:\anaconda\envs\weit_python\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [ ]:
model.summary()

Model: "패션 Mnist 모델"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ hidden (Dense)                  │ (None, 100)            │        78,500 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ output (Dense)                  │ (None, 10)             │         1,010 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 79,510 (310.59 KB)

 Trainable params: 79,510 (310.59 KB)

 Non-trainable params: 0 (0.00 B)

**방법3**

In [ ]:
model=keras.Sequential()
model.add(keras.layers.Dense(100, activation='sigmoid', input_shape=(784,)))
model.add(keras.layers.Dense(10, activation='softmax', name = 'output'))

In [ ]:
model.summary()

Model: "sequential_3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_2 (Dense)                 │ (None, 100)            │        78,500 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ output (Dense)                  │ (None, 10)             │         1,010 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 79,510 (310.59 KB)

 Trainable params: 79,510 (310.59 KB)

 Non-trainable params: 0 (0.00 B)

**모델 train**

In [ ]:
model.compile(loss = 'sparse_categorical_crossentropy', metrics=['accuracy'])
model.fit(train_scaled, train_target, epochs=5)

NameError: name 'model' is not defined

**RELU 함수**

시그모이드 함수: 오/왼쪽 끝으로 갈 수록 누워있는 그래프이다. 올바른 출력을 만들기 어려울 수도 있다.

RELU 함수: 0이하는 0, 0 이상은 y=x 함수, 이미지 처리에 좋은 성능을 냄

In [ ]:
# Flatten 층: 1차원으로 만들기

model=keras.Sequential()
model.add(keras.layers.Flatten(input_shape=(28, 28)))
model.add(keras.layers.Dense(100, activation='relu'))
model.add(keras.layers.Dense(10, activation='softmax'))

c:\anaconda\envs\weit_python\Lib\site-packages\keras\src\layers\reshaping\flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


In [ ]:
model.summary()
#Flatten층이 있다고 깊이가 3인 신경망이라고는 하지 않는다. (학습하는 층이 아니기 때문)

Model: "sequential_4"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ flatten (Flatten)               │ (None, 784)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ (None, 100)            │        78,500 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 10)             │         1,010 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 79,510 (310.59 KB)

 Trainable params: 79,510 (310.59 KB)

 Non-trainable params: 0 (0.00 B)

In [ ]:
(train_input, train_target), (test_input, test_target) = \
    keras.datasets.fashion_mnist.load_data()
train_scaled = train_input / 255.0
train_scaled, val_scaled, train_target, val_target = train_test_split(
    train_scaled, train_target, test_size=0.2, random_state=42)

In [ ]:
model.compile(loss = 'sparse_categorical_crossentropy', metrics=['accuracy'])
model.fit(train_scaled, train_target, epochs=5)

Epoch 1/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 2s 915us/step - accuracy: 0.7664 - loss: 0.6723
Epoch 2/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 1s 898us/step - accuracy: 0.8576 - loss: 0.3994
Epoch 3/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 1s 892us/step - accuracy: 0.8692 - loss: 0.3650
Epoch 4/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 1s 869us/step - accuracy: 0.8810 - loss: 0.3312
Epoch 5/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 2s 995us/step - accuracy: 0.8839 - loss: 0.3159


In [ ]:
model.evaluate(val_scaled, val_target)

375/375 ━━━━━━━━━━━━━━━━━━━━ 0s 735us/step - accuracy: 0.8739 - loss: 0.3637


[0.36868995428085327, 0.8726666569709778]

**optimizer**

compile 메서드에서 keras의 기본 경사 하강법 알고리즘인 RMSprop을 사용했다. keras는 다양한 경사 하강법 알고리즘을 제공하는데 이를 옵티마이저라고 한다.

In [ ]:
model.compile(optimizer='sgd', loss='sparse_categorical_entropy', metrics=['accuracy']) # SGD을 optimizer로 사용함
# 'sgd'는 sgd=keras.optimizers.SGD()와 동일함
# 학습률 조정: sgd=keras.optimizers.SGD(0.1)

<img src="optimizer.jpg" alt="description" width="500"/>

기본 경사 하강법 optimzer는 SGD class에서 제공한다. 

SGD 클래스의 momentum 매개변수의 기본 값은 0이다. 0보다 큰 값으로 지정하면 마치 이전의 gradient를 가속도처럼 사용하는 momentum optimzer를 사용한다.

SGD 클래스의 nestrov 매개변수를 기본 값 False에서 True로 바꾸면 Nestrov momentum optimizer를 사용한다.

대부분 nestrov가 momentum보다 더 나은 성능을 제공한다.

In [ ]:
sgd=keras.optimizers.SGD(momentum=0.9, nesterov=True)


모델이 최적점에 가까이 갈수록 학습률을 낮추는데 최적점에 수렴하는 학습률을 '적응적 학습률'이라고 한다. -> Adagrad와 RMSProp가 사용한다. 

In [ ]:
adagrad = keras.optimizers.Adagrad()
model.compile(optimizer=adagrad, loss='sparse_categorical_crossentropy', metrics=['accuracy'])

In [ ]:
rmsprop = keras.optimizers.RMSprop()
model.compile(optimizer=rmsprop, loss='sparse_categorical_crossentropy', metrics=['accuracy'])

momentum 최적화 + RMSProp의 장점을 접목한 것이 adam이다. 

In [ ]:
model=keras.Sequential()
model.add(keras.layers.Flatten(input_shape=(28, 28)))
model.add(keras.layers.Dense(100, activation='relu'))
model.add(keras.layers.Dense(10, activation='softmax'))

c:\anaconda\envs\weit_python\Lib\site-packages\keras\src\layers\reshaping\flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


In [ ]:
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model.fit(train_scaled, train_target, epochs=5)

Epoch 1/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.7646 - loss: 0.6703
Epoch 2/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 1s 904us/step - accuracy: 0.8548 - loss: 0.4092
Epoch 3/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 1s 965us/step - accuracy: 0.8693 - loss: 0.3606
Epoch 4/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.8803 - loss: 0.3311
Epoch 5/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.8867 - loss: 0.3089


In [ ]:
model.evaluate(val_scaled, val_target)

375/375 ━━━━━━━━━━━━━━━━━━━━ 0s 801us/step - accuracy: 0.8737 - loss: 0.3478


[0.34779882431030273, 0.8725000023841858]